coderd/wsconncache/wsconncache.go

```diff
@@ -0,0 +1,130 @@
+// Package wsconncache caches workspace agent connections by UUID.
+package wsconncache
+
+import (
+	"context"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/google/uuid"
+	"go.uber.org/atomic"
+	"golang.org/x/xerrors"
+
+	"github.com/coder/coder/agent"
+)
+
+// New creates a new workspace connection cache that closes
+// connections after the inactive timeout provided.
+//
+// Agent connections are cached due to WebRTC negotiation
+// taking a few hundred milliseconds.
+func New(dialer Dialer, inactiveTimeout time.Duration) *Cache {
+	if inactiveTimeout == 0 {
+		inactiveTimeout = 5 * time.Minute
+	}
+	return &Cache{
+		conns:           make(map[uuid.UUID]*Conn),
+		dialer:          dialer,
+		inactiveTimeout: inactiveTimeout,
+	}
+}
+
+// Dialer creates a new agent connection by ID.
+type Dialer func(r *http.Request, id uuid.UUID) (*agent.Conn, error)
+
+// Conn wraps an agent connection with a reusable HTTP transport.
+type Conn struct {
+	*agent.Conn
+
+	locks         atomic.Uint64
```

Comment:
you increment and decrement this number, but never read it or use it for any conditional logic.

Reply:
Yup, this is all WIP.